btrdbextras/dq.py

```diff
@@ -0,0 +1,409 @@
+import re
+import warnings
+from tabulate import tabulate
+
+import btrdb
+from btrdb.stream import StreamSet, Stream
+from btrdb.utils.general import pointwidth
+from btrdb.utils.timez import ns_delta, to_nanoseconds
+
+KNOWN_DISTILLER_TYPES = ["repeats", "duplicate-times", "zeros"]
+
+class AmbiguousDistillateError(ValueError):
+    """
+    Raised when a distillate stream's name contains references to multiple known
+    data quality distiller names
+    """
+    pass
+
+class DQDistillate(Stream):
+    """
+    Subsets a Stream object and allows for identfication of data quality events
+
+    Parameters
+    ----------
+    btrdb : BTrDB
+        A reference to the BTrDB object connecting this stream back to the
+		physical server.
+    uuid : UUID
+        The unique UUID identifier for this stream.
+    """
+    def __init__(self, btrdb, uu):
+        # gives all same attrs/methods as Stream
+        super().__init__(btrdb, uu)
+
+        # NOTE: this involves determining distiller type based on the distillate 
+        # stream name, so we will need to be careful how we name distillates
+        types = re.findall(r"(?=("+'|'.join(KNOWN_DISTILLER_TYPES)+r"))", self.name)
+        if len(types) == 0:
+            raise ValueError(f"unknown distiller type. Must be one of [{', '.join(KNOWN_DISTILLER_TYPES)}]")
+        if len(types) > 1:
+            raise AmbiguousDistillateError(f"ambiguous distiller name. contains references to [{', '.join(types)}]")
```

Comment:
I'm not 100% sure of what's going on here. What kinds of cases would raise this error?

Reply:
I'm trying to match the type of distiller that produced the distillate stream based on the distillate stream's name. An example stream name that our distillers produce would be like `PH0MAG L1MagAng-duplicate-times`. I created the `AmbiguousDistillateError` to deal with cases where the name is like `PH0MAG L1MagAng-duplicate-times-repeats`, where it contains references to more than one distiller type.

For now, we have full control over our distillers and the names of their output streams, I just wanted to include it here to catch it because it's not technically impossible, as users can change names of streams to whatever they want

Reply:
Ah I see, sounds good then.